plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/BaseJavaScriptVariantHelper.java

```diff
@@ -608,10 +629,41 @@ public String variantMode(EObject eobj) {
 	}
 
 	/**
-	 * Return true if "use strict" is declared. Override this method for sub-languages!
+	 * Returns true if "use strict" is declared. Override this method for sub-languages!
 	 */
 	@Override
 	public boolean isStrictMode(EObject eobj) {
 		return JavaScriptVariant.isContainedInStrictFunctionOrScript(eobj);
 	}
+
+	/**
+	 * Returns {@code true} if the script allows for distinct elements with the same qualified name in a scope.
+	 */
+	@Override
+	public boolean isMultiQNScope(EObject eobj) {
+		return get(fileExtensionCalculator.getXpectAwareFileExtension(eobj), MULTI_QN_SCOPE);
+	}
+
+	/**
+	 * Returns {@code true} if the script allows for the declaration and reference of versioned types as well as
+	 * corresponding migrations.
+	 */
+	@Override
+	public boolean allowVersionedTypes(EObject eobj) {
+		return get(fileExtensionCalculator.getXpectAwareFileExtension(eobj), VERSIONED_TYPES);
+	}
+
+	/**
+	 * Returns {@code true} if the script allows for top-level statements as opposed to just type and function
+	 * declarations.
+	 */
+	@Override
+	public boolean allowTopLevelStatements(EObject eobj) {
+		return get(fileExtensionCalculator.getXpectAwareFileExtension(eobj), TOP_LEVEL_STATEMENTS);
+	}
+
+	@Override
+	public String getVariantName(EObject eobj) {
+		return get(fileExtensionCalculator.getXpectAwareFileExtension(eobj), VARIANT_NAME);
+	}
```

Comment:
not sure if the distinction between `N4JSJavaScriptVariantHelper` and `BaseJavaScriptVariantHelper` and interface `JavaScriptVariantHelper` still makes sense after N4JSX and N4IDL have been merged into N4JS.

Please double check this and discuss with interested parties.

Reply:
I will discuss this with interested parties and create a follow-up issues.